### DISCUSSION TITLE
Rule validation

### DISCUSSION THREAD

**Discussion by Tresor-Kasenda** [livewire/livewire#6350] [1. Help] [answered]
#[Rule([
        'required',
        'integer',
        'exists:categories,id'
    ])]

 i would know how add a additional rule to the livewire validation system `Rule::unique('category', 'id')->ignore($this->id) `

**Comment by N1ebieski** [comment]
You can't. Use the old, good rules method

**Reply by Tresor-Kasenda** [reply]
**❔ question:** ‏
but if I want to update data and use ignore  #authenticated user information, how can I do that ???

**Reply by Kenneth-Inkum** [reply]
I've been trying to use either 

> Laravel\Fortify\Rules\Password;

 So I could do something like:
`['required', 'string', new Password, 'confirmed']` 

Or 

> Illuminate\Validation\Rules\Password;

So I could do something like:
`Password::defaults()`

But it just seems impossible. Meanwhile under the hood the new Livewire Rule Attribute uses 

> Illuminate\Support\Facades\Validator

**Reply by N1ebieski** [reply]
As i said, use the rules method in the form, something like this:

```php
public function rules(): array
{
    return [
        'name' => [
            'required',
            'string',
            'max:255',
            Rule::unique('users', 'name')->ignore($this->component->id)
        ],
   ];
}
```

Attribute Rule is useless at the moment

**Reply by sayedsadat344** [reply, marked as answer]
@N1ebieski 

Where should we define this rules array ?

Because we have an ` extracted form` which is outside the `component.php`, how can` $this->component->id` work here since it is not declared in this form?

We have defined the form as follow

### ProductForm.php

  ```
#[Rule('required|unique:products,name|min:5')]
    public $name = '';

    #[Rule('required|min:10')]
    public $description = '';

    #[Rule('required|numeric|gt:0')]
    public $price = 0.0;

    #[Rule('required|numeric|gt:0')]
    public $stock_quantity = 0;
```

### Product.php

```

   public $product_id;
    public ProductForm $productForm;

   public function updateProduct()
    {

        $this->validate();
        $prod->update($this->productForm->all());
        $this->dispatch('alert', [
            'type' => 'success',
            'message' => "Product updated Successfully!!",
        ]);

        return $this->redirect('/products', navigate: true);

    }
```

**Reply by sayedsadat344** [reply]
I found out that the newly added `Rule attribute` i.e. `#[Rule('required|unique:products,name|min:5')]` is not capable to perform `ignore` functionality during validation.

The solution would be to use old format of the `Rules Array` to overwrite the rule

```
 public function rules(): array
    {
        return [
            'productForm.name' => [
                'required',
                'string',
                'max:255',
                'min:5',
                Rule::unique('products', 'name')->ignore($this->product_id),
            ],
        ];
    }
```

Although it works, but at the same time, It indicates that the `V3 #Rule format` is not usable for complex validations.

**Reply by Tresor-Kasenda** [reply]
But in livewire form it's impossible to use ignore method ?

**Reply by sayedsadat344** [reply]
@Tresor-Kasenda Yep, there is no way to use ignore with form in livewire when you use `V3 #Rule format`. However, you can have an `extracted form object & specify the rules array in the component file`.

ex.

The form object `productForm`

 ```
 public $name = '';
  public $description = '';
  public $price = 0.0;
  public $stock_quantity = 0;
  
```
  component file `Product.php`
  
 ```
  public function rules(): array
    {
        return [
            'productForm.name' => [
                'required',
                'string',
                'max:255',
                'min:5',
                Rule::unique('products', 'name')->ignore($this->product_id),
            ],
           'productForm.description ' => [
            'required',
                'string',
                'max:255',
                'min:5',
           ],
          //other properties

        ];
    }
```

**Reply by Tresor-Kasenda** [reply]
okay thanks everybody for your answers

**Reply by krekas** [reply]
Believe it or on it is possible to use ignore in the form object. You can get value from the component. In you component you have a mount method something like so.
```php
public ?Product $product = null;
public Forms\ProductForm $form;

public function mount(Product $product = null): void
{
    if ($product->exists) {
        $this->form->setProduct($product);
    }
}
```

Then in the rules, which is a form object, we do:
```php
public function rules(): array
{
    return [
        'name'        => [
            'required',
            Rule::unique('products', 'name')->ignore($this->component->product)
        ],
    ];
}
```
